File: dart/test/semantic_index_runtime_observation_routes_test.dart
// FUTURE-PARITY-BACKLOG.10.5.5.3 — generated and emitted observation routes.

import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_dart/src/semantic/sha256.dart';
import 'package:test/test.dart';

const _input = 'ab\n';
const _responseDigest =
    '36897041c6f71b95b577ce7b38f42d3649c6adffc6c37c069944a90f6eb65887';
const _identity = 'semantic-introspection/runtime.spec';

void main() {
  late String source;
  late CompiledSpec compiled;
  late List<GeneratedPlanRow> plan;
  late SemanticIndex base;

  setUpAll(() {
    source = File(
      '../capability_conformance/semantic_introspection/runtime.spec',
    ).readAsStringSync();
    final parsed = parseSpec(source);
    validateSpec(parsed);
    compiled = compileSpec(parsed);
    plan = buildGeneratedRulePlan(compiled);
    base = SemanticIndex.fromSource(
      source,
      options: const SemanticIndexOptions(
        logicalName: 'runtime.spec',
        sourceDetailCeiling: SemanticSourceDetail.text,
      ),
    );
  });

  test('public generated helpers preserve exact observations and outputs', () {
    final baselineDiagnostics = <RuntimeDiagnosticOutputEvent>[];
    final baseline = executeGeneratedParserV2(
      compiled,
      plan,
      _input,
      _identity,
      diagnosticOutputSink: baselineDiagnostics.add,
    );
    final events = <RuntimeSemanticObservationEvent>[];
    final observedDiagnostics = <RuntimeDiagnosticOutputEvent>[];
    final observed = executeGeneratedParserV2(
      compiled,
      plan,
      _input,
      _identity,
      diagnosticOutputSink: observedDiagnostics.add,
      semanticObservationSink: events.add,
    );

    expect(observed, baseline);
    expect(_eventJson(events), _eventJson(_expectedEvents));
    expect(
      _diagnosticJson(observedDiagnostics),
      _diagnosticJson(baselineDiagnostics),
    );
    _expectTwentiethDigest(base, events);

    final scratch = Directory.systemTemp.createTempSync(
      'linkedspec-dart-semantic-generated-trace-',
    );
    try {
      final baselineTrace = '${scratch.path}/baseline.trace';
      final observedTrace = '${scratch.path}/observed.trace';
      final tracedBaseline = executeGeneratedParserWithTraceV2(
        compiled,
        plan,
        _input,
        LinkedSpecTraceConfig(
          level: LinkedSpecTraceLevel.debug,
          traceFile: baselineTrace,
          sinkMode: LinkedSpecTraceSinkMode.route,
          resetFile: true,
        ),
        _identity,
      );
      final tracedEvents = <RuntimeSemanticObservationEvent>[];
      final tracedObserved = executeGeneratedParserWithTraceV2(
        compiled,
        plan,
        _input,
        LinkedSpecTraceConfig(
          level: LinkedSpecTraceLevel.debug,
          traceFile: observedTrace,
          sinkMode: LinkedSpecTraceSinkMode.route,
          resetFile: true,
        ),
        _identity,
        semanticObservationSink: tracedEvents.add,
      );

      expect(tracedObserved, tracedBaseline);
      expect(
        File(observedTrace).readAsStringSync(),
        File(baselineTrace).readAsStringSync(),
      );
      expect(_eventJson(tracedEvents), _eventJson(_expectedEvents));
      _expectTwentiethDigest(base, tracedEvents);
    } finally {
      scratch.deleteSync(recursive: true);
    }

    for (final route in <void Function(RuntimeSemanticObservationSink sink)>[
      (sink) => executeGeneratedParserV2(
        compiled,
        plan,
        _input,
        _identity,
        semanticObservationSink: sink,
      ),
      (sink) => executeGeneratedParserWithTraceV2(
        compiled,
        plan,
        _input,
        LinkedSpecTraceConfig.disabled(),
        _identity,
        semanticObservationSink: sink,
      ),
    ]) {
      final failure = _CallerObservationFailure();
      final failureStack = StackTrace.current;
      try {
        route((_) => Error.throwWithStackTrace(failure, failureStack));
        fail('generated observation sink returned normally');
      } on Object catch (error, stackTrace) {
        expect(identical(error, failure), isTrue);
        expect(identical(stackTrace, failureStack), isTrue);
      }
    }

    final exitCompiled = _compile('''
Top::
 /x/
 E { exit_now(7) }
''');
    final exitEvents = <RuntimeSemanticObservationEvent>[];
    expect(
      () => executeGeneratedParserV2(
        exitCompiled,
        buildGeneratedRulePlan(exitCompiled),
        'x',
        'semantic-introspection/exit.spec',
        semanticObservationSink: exitEvents.add,
      ),
      throwsA(
        isA<RuntimeExitNow>().having((error) => error.status, 'status', 7),
      ),
    );
    expect(exitEvents, hasLength(1));
    expect(
      exitEvents.single.eventKind,
      RuntimeSemanticObservationEventKind.regexSlotSelected,
    );
  });

  test(
    'standalone emitted direct and traced APIs preserve exact observations',
    () async {
      final scratch = Directory.systemTemp.createTempSync(
        'linkedspec-dart-semantic-emitted-',
      );
      final pubCache = Directory('${scratch.path}/pub-cache')..createSync();
      final packageRoot = Directory.current.absolute;
      try {
        Directory('${scratch.path}/lib').createSync();
        Directory('${scratch.path}/bin').createSync();
        File('${scratch.path}/pubspec.yaml').writeAsStringSync('''
name: linkedspec_semantic_emitted_probe
publish_to: none
environment:
  sdk: ">=3.9.0 <4.0.0"
dependencies:
  linkedspec_dart:
    path: ${jsonEncode(packageRoot.path)}
''');
        File(
          '${scratch.path}/lib/generated.dart',
        ).writeAsStringSync(emitDartSourceV2(compiled, _identity));
        final exitCompiled = _compile('''
Top::
 /x/
 E { exit_now(7) }
''');
        File('${scratch.path}/lib/exit_generated.dart').writeAsStringSync(
          emitDartSourceV2(
            exitCompiled,
            'semantic-introspection/emitted-exit.spec',
          ),
        );
        File('${scratch.path}/bin/main.dart').writeAsStringSync(r'''
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_semantic_emitted_probe/exit_generated.dart'
    as exit_generated;
import 'package:linkedspec_semantic_emitted_probe/generated.dart' as generated;

final class CallerObservationFailure implements Exception {}

bool preservesFailure(
  void Function(RuntimeSemanticObservationSink sink) invoke,
) {
  final failure = CallerObservationFailure();
  final failureStack = StackTrace.current;
  try {
    invoke((_) => Error.throwWithStackTrace(failure, failureStack));
  } on Object catch (error, stackTrace) {
    return identical(error, failure) && identical(stackTrace, failureStack);
  }
  return false;
}

void main() {
  const input = 'ab\n';
  final directEvents = <RuntimeSemanticObservationEvent>[];
  final baselineDiagnostics = <RuntimeDiagnosticOutputEvent>[];
  final observedDiagnostics = <RuntimeDiagnosticOutputEvent>[];
  final baseline = generated.execute(
    input,
    diagnosticOutputSink: baselineDiagnostics.add,
  );
  final observed = generated.execute(
    input,
    diagnosticOutputSink: observedDiagnostics.add,
    semanticObservationSink: directEvents.add,
  );

  final tracedEvents = <RuntimeSemanticObservationEvent>[];
  final tracedBaseline = generated.executeWithTrace(
    input,
    const LinkedSpecTraceConfig(
      level: LinkedSpecTraceLevel.debug,
      traceFile: 'baseline.trace',
      sinkMode: LinkedSpecTraceSinkMode.route,
      resetFile: true,
    ),
  );
  final tracedObserved = generated.executeWithTrace(
    input,
    const LinkedSpecTraceConfig(
      level: LinkedSpecTraceLevel.debug,
      traceFile: 'observed.trace',
      sinkMode: LinkedSpecTraceSinkMode.route,
      resetFile: true,
    ),
    semanticObservationSink: tracedEvents.add,
  );

  final directFailureIdentity = preservesFailure(
    (sink) => generated.execute(input, semanticObservationSink: sink),
  );
  final tracedFailureIdentity = preservesFailure(
    (sink) => generated.executeWithTrace(
      input,
      LinkedSpecTraceConfig.disabled(),
      semanticObservationSink: sink,
    ),
  );

  final exitEvents = <RuntimeSemanticObservationEvent>[];
  int? exitStatus;
  try {
    exit_generated.execute('x', semanticObservationSink: exitEvents.add);
  } on RuntimeExitNow catch (error) {
    exitStatus = error.status;
  }

  print(jsonEncode({
    'contract': generated.linkedspecGeneratedSourceContract,
    'format': generated.linkedspecGeneratedSourceFormat,
    'baseline': baseline,
    'observed': observed,
    'traced_baseline': tracedBaseline,
    'traced_observed': tracedObserved,
    'direct_events': [for (final event in directEvents) event.toJson()],
    'traced_events': [for (final event in tracedEvents) event.toJson()],
    'diagnostics_equal': jsonEncode([
      for (final event in baselineDiagnostics) event.toJson(),
    ]) == jsonEncode([
      for (final event in observedDiagnostics) event.toJson(),
    ]),
    'trace_equal': File('baseline.trace').readAsStringSync() ==
        File('observed.trace').readAsStringSync(),
    'direct_failure_identity': directFailureIdentity,
    'traced_failure_identity': tracedFailureIdentity,
    'exit_status': exitStatus,
    'exit_events': [for (final event in exitEvents) event.toJson()],
  }));
}
''');

        final environment = {
          ...Platform.environment,
          'PUB_CACHE': pubCache.path,
        };
        await _expectProcessSuccess(scratch, environment, const [
          'pub',
          'get',
          '--offline',
        ]);
        await _expectProcessSuccess(scratch, environment, const [
          'analyze',
          '--fatal-infos',
          '--fatal-warnings',
        ]);
        final run = await _expectProcessSuccess(scratch, environment, const [
          'run',
          'bin/main.dart',
        ]);
        final output = (jsonDecode((run.stdout as String).trim()) as Map)
            .cast<String, Object?>();

        expect(output['contract'], linkedSpecGeneratedSourceContract);
        expect(output['format'], linkedSpecGeneratedSourceFormatVersion);
        expect(output['observed'], output['baseline']);
        expect(output['traced_observed'], output['traced_baseline']);
        expect(output['diagnostics_equal'], isTrue);
        expect(output['trace_equal'], isTrue);
        expect(output['direct_failure_identity'], isTrue);
        expect(output['traced_failure_identity'], isTrue);
        expect(output['exit_status'], 7);

        final directEvents = _eventsFromJson(output['direct_events']);
        final tracedEvents = _eventsFromJson(output['traced_events']);
        expect(_eventJson(directEvents), _eventJson(_expectedEvents));
        expect(_eventJson(tracedEvents), _eventJson(_expectedEvents));
        _expectTwentiethDigest(base, directEvents);
        _expectTwentiethDigest(base, tracedEvents);

        final exitEvents = _eventsFromJson(output['exit_events']);
        expect(exitEvents, hasLength(1));
        expect(
          exitEvents.single.eventKind,
          RuntimeSemanticObservationEventKind.regexSlotSelected,
        );
      } finally {
        scratch.deleteSync(recursive: true);
      }
      expect(scratch.existsSync(), isFalse);
    },
    timeout: const Timeout(Duration(minutes: 3)),
  );
}

CompiledSpec _compile(String source) {
  final parsed = parseSpec(source);
  validateSpec(parsed);
  return compileSpec(parsed);
}

List<Map<String, Object?>> _eventJson(
  List<RuntimeSemanticObservationEvent> events,
) => [for (final event in events) event.toJson()];

List<Map<String, Object?>> _diagnosticJson(
  List<RuntimeDiagnosticOutputEvent> events,
) => [for (final event in events) event.toJson()];

List<RuntimeSemanticObservationEvent> get _expectedEvents => [
  RuntimeSemanticObservationEvent.regexSlotSelected(
    ruleLabel: 'Top',
    targetRule: 'Top',
    regexIndex: 0,
    position: 1,
  ),
  RuntimeSemanticObservationEvent.regexSlotSelected(
    ruleLabel: 'Top',
    targetRule: 'Top',
    regexIndex: 1,
    position: 2,
  ),
  RuntimeSemanticObservationEvent.ruleResult(
    ruleLabel: 'Top',
    position: 2,
    input: _input,
  ),
];

List<RuntimeSemanticObservationEvent> _eventsFromJson(Object? value) => [
  for (final raw in (value! as List<Object?>))
    _eventFromJson((raw! as Map).cast<String, Object?>()),
];

RuntimeSemanticObservationEvent _eventFromJson(Map<String, Object?> value) =>
    RuntimeSemanticObservationEvent(
      contractId: value['contract_id']! as String,
      eventKind: RuntimeSemanticObservationEventKind.values.singleWhere(
        (kind) => kind.wireName == value['event_kind'],
      ),
      ruleLabel: value['rule_label']! as String,
      targetRule: value['target_rule'] as String?,
      regexIndex: value['regex_index'] as int?,
      position: value['position']! as int,
      inputIdentity: value['input_identity'] as String?,
      status: value['status'] as String?,
    );

void _expectTwentiethDigest(
  SemanticIndex base,
  List<RuntimeSemanticObservationEvent> events,
) {
  final response = base
      .withExecutionObservation(events)
      .query(
        SemanticQuery(
          operation: SemanticQueryOperation.list,
          recordKinds: const ['execution', 'event'],
          source: const SemanticQuerySource(
            detail: SemanticSourceDetail.identity,
          ),
        ),
      );
  expect(_canonicalDigest(response.toJson()), _responseDigest);
}

String _canonicalDigest(Map<String, Object?> value) =>
    sha256Hex(utf8.encode(jsonEncode(_canonicalValue(value))));

Object? _canonicalValue(Object? value) {
  if (value case Map<Object?, Object?>()) {
    final keys = value.keys.map((key) => key.toString()).toList()..sort();
    return <String, Object?>{
      for (final key in keys) key: _canonicalValue(value[key]),
    };
  }
  if (value case List<Object?>()) {
    return [for (final item in value) _canonicalValue(item)];
  }
  return value;
}

Future<ProcessResult> _expectProcessSuccess(
  Directory workingDirectory,
  Map<String, String> environment,
  List<String> arguments,
) async {
  final result = await Process.run(
    Platform.resolvedExecutable,
    arguments,
    workingDirectory: workingDirectory.path,
    environment: environment,
  );
  expect(
    result.exitCode,
    0,
    reason:
        'dart ${arguments.join(' ')} failed\n'
        'stdout:\n${result.stdout}\n'
        'stderr:\n${result.stderr}',
  );
  return result;
}

final class _CallerObservationFailure implements Exception {}

File: julia/test/semantic_index_runtime_observation_routes_test.jl
# FUTURE-PARITY-BACKLOG.10.6.6.3 — generated and emitted observation routes.

const JULIA_RUNTIME_ROUTE_RESPONSE_DIGEST =
    "36897041c6f71b95b577ce7b38f42d3649c6adffc6c37c069944a90f6eb65887"
const JULIA_RUNTIME_ROUTE_IDENTITY = "semantic-introspection/runtime.spec"

struct JuliaRuntimeRouteCallerFailure <: Exception
    id::String
end

function _julia_runtime_route_source()
    return read(
        joinpath(
            REPO_ROOT,
            "capability_conformance",
            "semantic_introspection",
            "runtime.spec",
        ),
        String,
    )
end

function _julia_runtime_route_base(source)
    return semantic_index(
        source;
        logical_name = "runtime.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )
end

function _julia_runtime_route_request()
    return SemanticQuery(
        operation = SemanticQueryListOperation,
        record_kinds = ("execution", "event"),
        source = SemanticQuerySource(detail = SemanticSourceIdentityDetail),
    )
end

function _julia_runtime_route_neutral_request()
    return Dict{String,Any}(
        "contract" => "linkedspec-semantic-query-v1",
        "operation" => "list",
        "subjects" => Any[],
        "record_kinds" => Any["execution", "event"],
        "relation_kinds" => Any[],
        "direction" => "outgoing",
        "page" => Dict{String,Any}("after_id" => nothing, "limit" => 100),
        "budget" => Dict{String,Any}(
            "max_records" => 1000,
            "max_relations" => 2000,
            "max_depth" => 4,
        ),
        "source" => Dict{String,Any}(
            "detail" => "identity",
            "include_content_digest" => false,
        ),
    )
end

function _julia_runtime_route_check_digest(base, events)
    observed = with_execution_observation(base, events)
    typed = semantic_query(observed, _julia_runtime_route_request())
    neutral = semantic_query_neutral(
        observed,
        _julia_runtime_route_neutral_request(),
    )
    @test typed == neutral
    @test _semantic_query_kernel_digest(typed) == JULIA_RUNTIME_ROUTE_RESPONSE_DIGEST
    return nothing
end

function _julia_runtime_route_capture(call)
    events = RuntimeSemanticObservationEvent[]
    value = call(event -> push!(events, event))
    return value, events
end

function _julia_runtime_route_event_from_json(raw)
    kind = raw["event_kind"] == "regex_slot_selected" ?
        RuntimeSemanticRegexSlotSelected : RuntimeSemanticRuleResult
    return RuntimeSemanticObservationEvent(
        contract_id = raw["contract_id"],
        event_kind = kind,
        rule_label = raw["rule_label"],
        target_rule = raw["target_rule"],
        regex_index = raw["regex_index"],
        position = raw["position"],
        input_identity = raw["input_identity"],
        status = raw["status"],
    )
end

function _julia_runtime_route_host_process(
    scratch,
    runner,
    generated,
    exit_generated,
    private_depot,
)
    separator = Sys.iswindows() ? ';' : ':'
    parent_depot = get(ENV, "JULIA_DEPOT_PATH", "")
    depot_path = isempty(parent_depot) ?
        private_depot : string(private_depot, separator, parent_depot)
    load_path = join(
        [scratch, joinpath(REPO_ROOT, "julia"), "@stdlib"],
        separator,
    )
    command = `$(Base.julia_cmd()) --project=$scratch --startup-file=no --history-file=no --compiled-modules=no $runner $generated $exit_generated`
    environment = copy(ENV)
    environment["JULIA_DEPOT_PATH"] = depot_path
    environment["JULIA_LOAD_PATH"] = load_path
    environment["JULIA_PKG_OFFLINE"] = "true"
    output = IOBuffer()
    process = run(
        pipeline(ignorestatus(setenv(command, environment)); stdout = output, stderr = output),
    )
    return success(process), String(take!(output))
end

@testset "Julia generated and emitted semantic observation routes" begin
    source = _julia_runtime_route_source()
    compiled = _julia_runtime_observation_compile(source)
    plan = build_generated_rule_plan(compiled)
    base = _julia_runtime_route_base(source)
    expected = _julia_runtime_observation_expected_events()

    @testset "public generated helpers preserve outputs trace diagnostics and identity" begin
        baseline_diagnostics = RuntimeDiagnosticOutputEvent[]
        baseline = execute_generated_parser_v2(
            compiled,
            plan,
            JULIA_RUNTIME_OBSERVATION_INPUT,
            JULIA_RUNTIME_ROUTE_IDENTITY;
            diagnostic_output_sink = event -> push!(baseline_diagnostics, event),
        )
        observed_diagnostics = RuntimeDiagnosticOutputEvent[]
        observed, events = _julia_runtime_route_capture() do sink
            execute_generated_parser_v2(
                compiled,
                plan,
                JULIA_RUNTIME_OBSERVATION_INPUT,
                JULIA_RUNTIME_ROUTE_IDENTITY;
                diagnostic_output_sink = event -> push!(observed_diagnostics, event),
                semantic_observation_sink = sink,
            )
        end
        @test observed == baseline == Any["A", "B"]
        @test events == expected
        @test [to_json(event) for event in observed_diagnostics] ==
              [to_json(event) for event in baseline_diagnostics]
        _julia_runtime_route_check_digest(base, events)

        baseline_trace = IOBuffer()
        traced_baseline = execute_generated_parser_with_trace_v2(
            compiled,
            plan,
            JULIA_RUNTIME_OBSERVATION_INPUT,
            trace_config_enabled(LinkedSpecTraceDebug),
            JULIA_RUNTIME_ROUTE_IDENTITY;
            stdout_io = baseline_trace,
        )
        observed_trace = IOBuffer()
        traced_observed, traced_events = _julia_runtime_route_capture() do sink
            execute_generated_parser_with_trace_v2(
                compiled,
                plan,
                JULIA_RUNTIME_OBSERVATION_INPUT,
                trace_config_enabled(LinkedSpecTraceDebug),
                JULIA_RUNTIME_ROUTE_IDENTITY;
                stdout_io = observed_trace,
                semantic_observation_sink = sink,
            )
        end
        @test traced_observed == traced_baseline == Any["A", "B"]
        @test String(take!(observed_trace)) == String(take!(baseline_trace))
        @test traced_events == expected
        _julia_runtime_route_check_digest(base, traced_events)

        routes = Function[
            sink -> execute_generated_parser_v2(
                compiled,
                plan,
                JULIA_RUNTIME_OBSERVATION_INPUT,
                JULIA_RUNTIME_ROUTE_IDENTITY;
                semantic_observation_sink = sink,
            ),
            sink -> execute_generated_parser_with_trace_v2(
                compiled,
                plan,
                JULIA_RUNTIME_OBSERVATION_INPUT,
                trace_config_disabled(),
                JULIA_RUNTIME_ROUTE_IDENTITY;
                semantic_observation_sink = sink,
            ),
        ]
        for (index, route) in enumerate(routes)
            failure = JuliaRuntimeRouteCallerFailure("generated-$index")
            captured = try
                route(_ -> throw(failure))
                nothing
            catch error
                error
            end
            @test captured === failure
        end

        exit_compiled = _julia_runtime_observation_compile(
            "Top::\n /x/\n E { exit_now(7) }\n",
        )
        exit_events = RuntimeSemanticObservationEvent[]
        captured_exit = try
            execute_generated_parser_v2(
                exit_compiled,
                build_generated_rule_plan(exit_compiled),
                "x",
                "semantic-introspection/exit.spec";
                semantic_observation_sink = event -> push!(exit_events, event),
            )
            nothing
        catch error
            error
        end
        @test captured_exit isa RuntimeExitNow
        @test captured_exit.status == 7
        @test length(exit_events) == 1
        @test only(exit_events).event_kind == RuntimeSemanticRegexSlotSelected
    end

    @testset "fresh emitted direct and traced wrappers forward the existing sink" begin
        generated = emit_julia_source_v2(compiled, JULIA_RUNTIME_ROUTE_IDENTITY)
        @test generated == emit_julia_source_v2(compiled, JULIA_RUNTIME_ROUTE_IDENTITY)
        @test length(findall("semantic_observation_sink = nothing", generated)) == 2
        @test length(findall(
            "semantic_observation_sink = semantic_observation_sink",
            generated,
        )) == 2
        @test occursin("const LINKEDSPEC_GENERATED_SOURCE_FORMAT = 2", generated)
        @test occursin(
            "const LINKEDSPEC_GENERATED_SOURCE_CONTRACT = \"linkedspec-generated-source-v2\"",
            generated,
        )

        mktempdir() do scratch
            generated_path = joinpath(scratch, "runtime_generated.jl")
            write(generated_path, generated)
            host = Module(:JuliaRuntimeObservationGeneratedHost)
            Base.include(host, generated_path)
            parser = Base.invokelatest(
                getproperty,
                host,
                :LinkedSpecGeneratedParser,
            )
            execute = Base.invokelatest(getproperty, parser, :execute)
            execute_with_trace = Base.invokelatest(
                getproperty,
                parser,
                :execute_with_trace,
            )

            baseline_diagnostics = RuntimeDiagnosticOutputEvent[]
            baseline = Base.invokelatest(
                execute,
                JULIA_RUNTIME_OBSERVATION_INPUT;
                diagnostic_output_sink = event -> push!(baseline_diagnostics, event),
            )
            observed_diagnostics = RuntimeDiagnosticOutputEvent[]
            observed, events = _julia_runtime_route_capture() do sink
                Base.invokelatest(
                    execute,
                    JULIA_RUNTIME_OBSERVATION_INPUT;
                    diagnostic_output_sink = event -> push!(observed_diagnostics, event),
                    semantic_observation_sink = sink,
                )
            end
            @test observed == baseline == Any["A", "B"]
            @test events == expected
            @test [to_json(event) for event in observed_diagnostics] ==
                  [to_json(event) for event in baseline_diagnostics]
            _julia_runtime_route_check_digest(base, events)

            baseline_trace = IOBuffer()
            traced_baseline = Base.invokelatest(
                execute_with_trace,
                JULIA_RUNTIME_OBSERVATION_INPUT,
                trace_config_enabled(LinkedSpecTraceDebug);
                stdout_io = baseline_trace,
            )
            observed_trace = IOBuffer()
            traced_observed, traced_events = _julia_runtime_route_capture() do sink
                Base.invokelatest(
                    execute_with_trace,
                    JULIA_RUNTIME_OBSERVATION_INPUT,
                    trace_config_enabled(LinkedSpecTraceDebug);
                    stdout_io = observed_trace,
                    semantic_observation_sink = sink,
                )
            end
            @test traced_observed == traced_baseline == Any["A", "B"]
            @test String(take!(observed_trace)) == String(take!(baseline_trace))
            @test traced_events == expected
            _julia_runtime_route_check_digest(base, traced_events)

            routes = Function[
                sink -> Base.invokelatest(
                    execute,
                    JULIA_RUNTIME_OBSERVATION_INPUT;
                    semantic_observation_sink = sink,
                ),
                sink -> Base.invokelatest(
                    execute_with_trace,
                    JULIA_RUNTIME_OBSERVATION_INPUT,
                    trace_config_disabled();
                    semantic_observation_sink = sink,
                ),
            ]
            for (index, route) in enumerate(routes)
                failure = JuliaRuntimeRouteCallerFailure("emitted-$index")
                captured = try
                    route(_ -> throw(failure))
                    nothing
                catch error
                    error
                end
                @test captured === failure
            end
        end
    end

    @testset "isolated emitted host preserves direct traced exit and callback behavior" begin
        mktempdir() do scratch
            private_depot = joinpath(scratch, "depot")
            mkpath(private_depot)
            write(
                joinpath(scratch, "Project.toml"),
                "name = \"JuliaSemanticObservationHost\"\n" *
                "uuid = \"bfa02ea0-b421-4f41-86d4-787799f1a3d1\"\n" *
                "version = \"0.1.0\"\n",
            )
            generated_path = joinpath(scratch, "runtime_generated.jl")
            exit_path = joinpath(scratch, "exit_generated.jl")
            runner_path = joinpath(scratch, "runner.jl")
            write(
                generated_path,
                emit_julia_source_v2(compiled, JULIA_RUNTIME_ROUTE_IDENTITY),
            )
            exit_compiled = _julia_runtime_observation_compile(
                "Top::\n /x/\n E { exit_now(7) }\n",
            )
            write(
                exit_path,
                emit_julia_source_v2(
                    exit_compiled,
                    "semantic-introspection/emitted-exit.spec",
                ),
            )
            write(
                runner_path,
                raw"""
import JSON3
import LinkedSpecJulia

include(ARGS[1])
const Parser = LinkedSpecGeneratedParser
const INPUT = "ab\n"

struct CallerObservationFailure <: Exception end

function preserves_failure(call)
    failure = CallerObservationFailure()
    captured = try
        call(_ -> throw(failure))
        nothing
    catch error
        error
    end
    return captured === failure
end

baseline_diagnostics = LinkedSpecJulia.RuntimeDiagnosticOutputEvent[]
baseline = Parser.execute(
    INPUT;
    diagnostic_output_sink = event -> push!(baseline_diagnostics, event),
)
events = LinkedSpecJulia.RuntimeSemanticObservationEvent[]
observed_diagnostics = LinkedSpecJulia.RuntimeDiagnosticOutputEvent[]
observed = Parser.execute(
    INPUT;
    diagnostic_output_sink = event -> push!(observed_diagnostics, event),
    semantic_observation_sink = event -> push!(events, event),
)

baseline_trace = IOBuffer()
traced_baseline = Parser.execute_with_trace(
    INPUT,
    LinkedSpecJulia.trace_config_enabled(LinkedSpecJulia.LinkedSpecTraceDebug);
    stdout_io = baseline_trace,
)
observed_trace = IOBuffer()
traced_events = LinkedSpecJulia.RuntimeSemanticObservationEvent[]
traced_observed = Parser.execute_with_trace(
    INPUT,
    LinkedSpecJulia.trace_config_enabled(LinkedSpecJulia.LinkedSpecTraceDebug);
    stdout_io = observed_trace,
    semantic_observation_sink = event -> push!(traced_events, event),
)

direct_failure = preserves_failure() do sink
    Parser.execute(INPUT; semantic_observation_sink = sink)
end
traced_failure = preserves_failure() do sink
    Parser.execute_with_trace(
        INPUT,
        LinkedSpecJulia.trace_config_disabled();
        semantic_observation_sink = sink,
    )
end

exit_host = Module(:JuliaSemanticObservationExitHost)
Base.include(exit_host, ARGS[2])
exit_parser = getfield(exit_host, :LinkedSpecGeneratedParser)
exit_events = LinkedSpecJulia.RuntimeSemanticObservationEvent[]
exit_failure = try
    exit_parser.execute(
        "x";
        semantic_observation_sink = event -> push!(exit_events, event),
    )
    nothing
catch error
    error
end

print(JSON3.write(Dict{String,Any}(
    "contract" => Parser.LINKEDSPEC_GENERATED_SOURCE_CONTRACT,
    "format" => Parser.LINKEDSPEC_GENERATED_SOURCE_FORMAT,
    "baseline" => baseline,
    "observed" => observed,
    "traced_baseline" => traced_baseline,
    "traced_observed" => traced_observed,
    "events" => [LinkedSpecJulia.to_json(event) for event in events],
    "traced_events" => [LinkedSpecJulia.to_json(event) for event in traced_events],
    "diagnostics_equal" => [LinkedSpecJulia.to_json(event) for event in observed_diagnostics] ==
        [LinkedSpecJulia.to_json(event) for event in baseline_diagnostics],
    "trace_equal" => String(take!(observed_trace)) == String(take!(baseline_trace)),
    "direct_failure" => direct_failure,
    "traced_failure" => traced_failure,
    "exit_status" => exit_failure isa LinkedSpecJulia.RuntimeExitNow ? exit_failure.status : nothing,
    "exit_events" => [LinkedSpecJulia.to_json(event) for event in exit_events],
)))
""",
            )

            passed, output = _julia_runtime_route_host_process(
                scratch,
                runner_path,
                generated_path,
                exit_path,
                private_depot,
            )
            @test passed
            payload = passed ? JSON3.read(output, Dict{String,Any}) : Dict{String,Any}()
            @test payload["contract"] == GENERATED_SOURCE_CONTRACT
            @test payload["format"] == GENERATED_SOURCE_FORMAT
            @test payload["observed"] == payload["baseline"] == Any["A", "B"]
            @test payload["traced_observed"] == payload["traced_baseline"] == Any["A", "B"]
            @test payload["diagnostics_equal"] === true
            @test payload["trace_equal"] === true
            @test payload["direct_failure"] === true
            @test payload["traced_failure"] === true
            @test payload["exit_status"] == 7

            direct_events = RuntimeSemanticObservationEvent[
                _julia_runtime_route_event_from_json(raw) for raw in payload["events"]
            ]
            traced_events = RuntimeSemanticObservationEvent[
                _julia_runtime_route_event_from_json(raw) for raw in payload["traced_events"]
            ]
            @test direct_events == expected
            @test traced_events == expected
            _julia_runtime_route_check_digest(base, direct_events)
            _julia_runtime_route_check_digest(base, traced_events)

            exit_events = RuntimeSemanticObservationEvent[
                _julia_runtime_route_event_from_json(raw) for raw in payload["exit_events"]
            ]
            @test length(exit_events) == 1
            @test only(exit_events).event_kind == RuntimeSemanticRegexSlotSelected
        end
    end
end
